File: SampleProject/Classes/Subclasses/UIButton/SPButton.m
//
//  SPButton.m
//  <#Project Name#>
//
//  Created by <#Project Developer#> on 12/11/2015.
//  Copyright (c) 2015 <#Project Name#>. All rights reserved.
//

#import "SPButton.h"

@implementation SPButton

- (void)commonInit
{
    self.iconSize = CGSizeZero;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (CGRect)titleRectForContentRect:(CGRect)contentRect
{
    
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated"
    CGSize size = [[self titleForState:self.state] sizeWithFont:self.font
                                              constrainedToSize:contentRect.size];
#pragma clang diagnostic pop
    CGSize iconSize = CGSizeEqualToSize(self.iconSize, CGSizeZero) ? [super imageRectForContentRect:contentRect].size : self.iconSize;
    CGFloat totalWidth = size.width + iconSize.width + self.iconMargin;
    CGFloat totalHeight = size.height + iconSize.height + self.iconMargin;
    CGRect rect = {{0, 0}, size};
    switch (self.contentHorizontalAlignment) {
        case UIControlContentHorizontalAlignmentLeft:
            switch (_iconPosition) {
                case SPIconPositionRight:
                    rect.origin.x = CGRectGetMinX(contentRect);
                    break;
                case SPIconPositionLeft:
                    rect.origin.x = CGRectGetMinX(contentRect) + totalWidth - size.width;
                    break;
                default:
                    rect.origin.x = CGRectGetMinX(contentRect);
                    break;
            }
            break;
        case UIControlContentHorizontalAlignmentRight:
            switch (_iconPosition) {
                case SPIconPositionRight:
                    rect.origin.x = CGRectGetMaxX(contentRect) - totalWidth;
                    break;
                case SPIconPositionLeft:
                    rect.origin.x = CGRectGetMaxX(contentRect) - size.width;
                    break;
                default:
                    rect.origin.x = CGRectGetMaxX(contentRect) - size.width;
                    break;
            }
            break;
        case UIControlContentHorizontalAlignmentFill:
            switch (_iconPosition) {
                case SPIconPositionRight:
                    rect.origin.x = CGRectGetMinX(contentRect);
                    break;
                case SPIconPositionLeft:
                    rect.origin.x = CGRectGetMaxX(contentRect) - size.width;
                    break;
                default:
                    rect.origin.x = CGRectGetMinX(contentRect) + (CGRectGetWidth(contentRect) - size.width) / 2;
                    break;
            }
            break;
        default:
            switch (_iconPosition) {
                case SPIconPositionRight:
                    rect.origin.x = CGRectGetMinX(contentRect) + (CGRectGetWidth(contentRect) - totalWidth) / 2;
                    break;
                case SPIconPositionLeft:
                    rect.origin.x = CGRectGetMinX(contentRect) + CGRectGetWidth(contentRect) - (CGRectGetWidth(contentRect) - totalWidth) / 2 - size.width;
                    break;
                default:
                    rect.origin.x = CGRectGetMinX(contentRect) + (CGRectGetWidth(contentRect) - size.width) / 2;
                    break;
            }
            break;
    }
    
    switch (self.contentVerticalAlignment) {
        case UIControlContentVerticalAlignmentTop:
            switch (_iconPosition) {
                case SPIconPositionTop:
                    rect.origin.y = CGRectGetMinY(contentRect) + totalHeight - size.height;
                    break;
                case SPIconPositionBottom:
                    rect.origin.y = CGRectGetMinY(contentRect);
                    break;
                default:
                    rect.origin.y = CGRectGetMinY(contentRect);
                    break;
            }
            break;
        case UIControlContentVerticalAlignmentBottom:
            switch (_iconPosition) {
                case SPIconPositionTop:
                    rect.origin.y = CGRectGetMaxY(contentRect) - size.height;
                    break;
                case SPIconPositionBottom:
                    rect.origin.y = CGRectGetMaxY(contentRect) - totalHeight;
                    break;
                default:
                    rect.origin.y = CGRectGetMaxY(contentRect) - size.height;
                    break;
            }
            break;
        case UIControlContentVerticalAlignmentFill:
            switch (_iconPosition) {
                case SPIconPositionTop:
                    rect.origin.y = CGRectGetMaxY(contentRect) - size.height;
                    break;
                case SPIconPositionBottom:
                    rect.origin.y = CGRectGetMinY(contentRect);
                    break;
                default:
                    rect.origin.y = CGRectGetMinY(contentRect) + (CGRectGetHeight(contentRect) - size.height) / 2;
                    break;
            }
            break;
        default:
            switch (_iconPosition) {
                case SPIconPositionTop:
                    rect.origin.y = CGRectGetMaxY(contentRect) - (CGRectGetHeight(contentRect) - totalHeight) / 2 - size.height;
                    break;
                case SPIconPositionBottom:
                    rect.origin.y = CGRectGetMinY(contentRect) + (CGRectGetHeight(contentRect) - totalHeight) / 2;
                    break;
                default:
                    rect.origin.y = CGRectGetMinY(contentRect) + (CGRectGetHeight(contentRect) - size.height) / 2;
                    break;
            }
            break;
    }
    return rect;
}

- (CGRect)imageRectForContentRect:(CGRect)contentRect
{
    CGSize size = CGSizeEqualToSize(self.iconSize, CGSizeZero) ? [super imageRectForContentRect:contentRect].size : self.iconSize;
    CGSize titleSize = [self titleRectForContentRect:contentRect].size;
    
    switch (_iconPosition) {
        case SPIconPositionTop:
        case SPIconPositionBottom:
            size.height = MAX(MIN(CGRectGetHeight(contentRect) - self.iconMargin - titleSize.height, size.height), self.iconSize.height);
            break;
        default:
            size.width = MAX(MIN(CGRectGetWidth(contentRect) - self.iconMargin - titleSize.width, size.width), self.iconSize.width);
            break;
    }
    
    CGFloat totalWidth = size.width + titleSize.width + self.iconMargin;
    CGFloat totalHeight = size.height + titleSize.height + self.iconMargin;
    CGRect rect = {{0, 0}, size};
    switch (self.contentHorizontalAlignment) {
        case UIControlContentHorizontalAlignmentLeft:
            switch (_iconPosition) {
                case SPIconPositionRight:
                    rect.origin.x = CGRectGetMinX(contentRect) + totalWidth - size.width;
                    break;
                case SPIconPositionLeft:
                    rect.origin.x = CGRectGetMinX(contentRect);
                    break;
                default:
                    rect.origin.x = CGRectGetMinX(contentRect);
                    break;
            }
            break;
        case UIControlContentHorizontalAlignmentRight:
            switch (_iconPosition) {
                case SPIconPositionRight:
                    rect.origin.x = CGRectGetMaxX(contentRect) - size.width;
                    break;
                case SPIconPositionLeft:
                    rect.origin.x = CGRectGetMaxX(contentRect) - totalWidth;
                    break;
                default:
                    rect.origin.x = CGRectGetMaxX(contentRect) - size.width;
                    break;
            }
            break;
        case UIControlContentHorizontalAlignmentFill:
            switch (_iconPosition) {
                case SPIconPositionRight:
                    rect.origin.x = CGRectGetMaxX(contentRect) - size.width;
                    break;
                case SPIconPositionLeft:
                    rect.origin.x = CGRectGetMinX(contentRect);
                    break;
                default:
                    rect.origin.x = CGRectGetMinX(contentRect) + (CGRectGetWidth(contentRect) - size.width) / 2;
                    break;
            }
            break;
        default:
            switch (_iconPosition) {
                case SPIconPositionRight:
                    rect.origin.x = CGRectGetMinX(contentRect) + CGRectGetWidth(contentRect) - (CGRectGetWidth(contentRect) - totalWidth) / 2 - size.width;
                    break;
                case SPIconPositionLeft:
                    rect.origin.x = CGRectGetMinX(contentRect) + (CGRectGetWidth(contentRect) - totalWidth) / 2;
                    break;
                default:
                    rect.origin.x = CGRectGetMinX(contentRect) + (CGRectGetWidth(contentRect) - size.width) / 2;
                    break;
            }
            break;
    }
    
    switch (self.contentVerticalAlignment) {
        case UIControlContentVerticalAlignmentTop:
            switch (_iconPosition) {
                case SPIconPositionTop:
                    rect.origin.y = CGRectGetMinY(contentRect);
                    break;
                case SPIconPositionBottom:
                    rect.origin.y = CGRectGetMinY(contentRect) + totalHeight - size.height;
                    break;
                default:
                    rect.origin.y = CGRectGetMinY(contentRect);
                    break;
            }
            break;
        case UIControlContentVerticalAlignmentBottom:
            switch (_iconPosition) {
                case SPIconPositionTop:
                    rect.origin.y = CGRectGetMaxY(contentRect) - totalHeight;
                    break;
                case SPIconPositionBottom:
                    rect.origin.y = CGRectGetMaxY(contentRect) - size.height;
                    break;
                default:
                    rect.origin.y = CGRectGetMaxY(contentRect) - size.height;
                    break;
            }
            break;
        case UIControlContentVerticalAlignmentFill:
            switch (_iconPosition) {
                case SPIconPositionTop:
                    rect.origin.y = CGRectGetMinY(contentRect);
                    break;
                case SPIconPositionBottom:
                    rect.origin.y = CGRectGetMaxY(contentRect) - size.height;
                    break;
                default:
                    rect.origin.y = CGRectGetMinY(contentRect) + (CGRectGetHeight(contentRect) - size.height) / 2;
                    break;
            }
            break;
        default:
            switch (_iconPosition) {
                case SPIconPositionTop:
                    rect.origin.y = CGRectGetMinY(contentRect) + (CGRectGetHeight(contentRect) - totalHeight) / 2;
                    break;
                case SPIconPositionBottom:
                    rect.origin.y = CGRectGetMaxY(contentRect) - (CGRectGetHeight(contentRect) - totalHeight) / 2 - size.height;
                    break;
                default:
                    rect.origin.y = CGRectGetMinY(contentRect) + (CGRectGetHeight(contentRect) - size.height) / 2;
                    break;
            }
            break;
    }
    
    return rect;
}

- (CGSize)intrinsicContentSize
{
    CGRect contentRect = [self contentRectForBounds:self.bounds];
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated"
    CGSize titleSize = [[self titleForState:self.state] sizeWithFont:self.font];
#pragma clang diagnostic pop
    CGSize imageSize = CGSizeEqualToSize(self.iconSize, CGSizeZero) ? [super imageRectForContentRect:contentRect].size : self.iconSize;
    
    switch (_iconPosition) {
        case SPIconPositionTop:
        case SPIconPositionBottom:
            return CGSizeMake(MAX(titleSize.width, imageSize.width), titleSize.height + imageSize.height + self.iconMargin);
            
            break;
        default:
            return CGSizeMake(titleSize.width + imageSize.width + self.iconMargin, MAX(titleSize.height, imageSize.height));
            break;
    }
}

- (CGSize)sizeThatFits:(CGSize)size
{
    return [self intrinsicContentSize];
}

@end

File: SampleProject/Classes/Subclasses/UIButton/SPButton.swift
//
//  SPButton.swift
//  <#Project Name#>
//
//  Created by <#Project Developer#> on 19/01/2017.
//  Copyright © 2017 <#Project Name#>. All rights reserved.
//

import Foundation
import UIKit

enum SPIconPosition: Int {
    case top = 0
    case left
    case bottom
    case right
}

@IBDesignable
class SPButton: UIButton {
    @IBInspectable var iconMargin: CGFloat = 0
    @IBInspectable var iconPositionRaw: Int {
        get { return iconPosition.rawValue }
        set { iconPosition = SPIconPosition(rawValue: newValue) ?? .left }
    }
    var iconPosition: SPIconPosition = .left
    @IBInspectable var iconSize: CGSize = .zero
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        commonInit()
    }
    
    private func commonInit() {
        iconSize = .zero
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        guard let imageView = imageView, let titleLabel = titleLabel else { return }
        
        let titleSize = titleLabel.intrinsicContentSize
        let imageSize = iconSize == .zero ? imageView.intrinsicContentSize : iconSize
        
        let totalWidth = titleSize.width + imageSize.width + iconMargin
        let totalHeight = titleSize.height + imageSize.height + iconMargin
        
        // Position the image and title based on the iconPosition
        switch iconPosition {
        case .top:
            imageView.frame = CGRect(
                x: (bounds.width - imageSize.width) / 2,
                y: (bounds.height - totalHeight) / 2,
                width: imageSize.width,
                height: imageSize.height
            )
            titleLabel.frame = CGRect(
                x: (bounds.width - titleSize.width) / 2,
                y: imageView.frame.maxY + iconMargin,
                width: titleSize.width,
                height: titleSize.height
            )
        case .bottom:
            titleLabel.frame = CGRect(
                x: (bounds.width - titleSize.width) / 2,
                y: (bounds.height - totalHeight) / 2,
                width: titleSize.width,
                height: titleSize.height
            )
            imageView.frame = CGRect(
                x: (bounds.width - imageSize.width) / 2,
                y: titleLabel.frame.maxY + iconMargin,
                width: imageSize.width,
                height: imageSize.height
            )
        case .left:
            imageView.frame = CGRect(
                x: (bounds.width - totalWidth) / 2,
                y: (bounds.height - imageSize.height) / 2,
                width: imageSize.width,
                height: imageSize.height
            )
            titleLabel.frame = CGRect(
                x: imageView.frame.maxX + iconMargin,
                y: (bounds.height - titleSize.height) / 2,
                width: titleSize.width,
                height: titleSize.height
            )
        case .right:
            titleLabel.frame = CGRect(
                x: (bounds.width - totalWidth) / 2,
                y: (bounds.height - titleSize.height) / 2,
                width: titleSize.width,
                height: titleSize.height
            )
            imageView.frame = CGRect(
                x: titleLabel.frame.maxX + iconMargin,
                y: (bounds.height - imageSize.height) / 2,
                width: imageSize.width,
                height: imageSize.height
            )
        }
    }
    
    override var intrinsicContentSize: CGSize {
        guard let imageView = imageView, let titleLabel = titleLabel else { return super.intrinsicContentSize }
        
        let titleSize = titleLabel.intrinsicContentSize
        let imageSize = iconSize == .zero ? imageView.intrinsicContentSize : iconSize
        
        switch iconPosition {
        case .top, .bottom:
            return CGSize(
                width: max(titleSize.width, imageSize.width),
                height: titleSize.height + imageSize.height + iconMargin
            )
        case .left, .right:
            return CGSize(
                width: titleSize.width + imageSize.width + iconMargin,
                height: max(titleSize.height, imageSize.height)
            )
        }
    }
    
    override func sizeThatFits(_ size: CGSize) -> CGSize {
        return intrinsicContentSize
    }
}
